### PR TITLE
Does source expression subsume another source expression?

### PR DESCRIPTION
**Pull request by Sun77789**
@mikewest

### CODE REVIEW

File: embedded/index.src.html [hunk 1]
```diff
@@ -414,6 +414,45 @@ <h3 id="subsume-policy" algorithm>
...
+  <h4 id="subsume-se" algorithm>
+    Does |scheme-source / host-source expression| subsume another 
+    |scheme-source / host-source expression|?
+  </h4>
+
+  Given two <a grammar>scheme-source</a> / <a grammar>host-source</a> 
+  expressions A and B, this algorithm returns "`Subsumes`" if A matches B, and 
+  returns "`Does Not Subsume`" otherwise.
+
+  Let <a grammar>`host-part`</a> be a `wildcard-host` if its first character is
```
Comment (Sun77789) [inline embedded/index.src.html]: Not sure if I should define these somewhere separately?

**Comment by Sun77789** [comment]
@mikewest updated!

File: embedded/index.src.html [hunk 2]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
     3.  Return "`Not Subsumed`".
   </ol>
 
+  <h4 id="subsume-se" algorithm>
+    Does |scheme-source / host-source expression| subsume another 
+    |scheme-source / host-source expression|?
```
Comment (mikewest) [inline embedded/index.src.html]: I'd rename this to "Does `source-expression` |A| subsume `source-expression` |B|?"

In the algorithm you can branch on the type of the expression, and just add an "ISSUE: TODO." for anything that doesn't match the `scheme-source` or `host-source` grammar.

File: embedded/index.src.html [hunk 3]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
...
+
+      1.  If |A| matches <a grammar>scheme-source</a>, return "`Subsumes`".
+          Otherwise, return "`Does Not Subsume`".
+
+  3.  If |B|'s <a grammar>`host-part`</a> is a `wildcard-host`:
+
+      1.  If |A|'s <a grammar>`host-part`</a> is not a `wildcard-host`,
+          return "`Does not Subsume`".
+
+      1.  Let |remaining| be the result of removing the leading ("*") from |B|'s
```
Comment (mikewest) [inline embedded/index.src.html]: Do you want to include the `.` after the `*` here?

File: embedded/index.src.html [hunk 4]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
...
+    Does |scheme-source / host-source expression| subsume another 
+    |scheme-source / host-source expression|?
+  </h4>
+
+  Given two <a grammar>scheme-source</a> / <a grammar>host-source</a> 
+  expressions |A| and |B|, this algorithm returns "`Subsumes`" if |A| subsumes
+  |B|, and returns "`Does Not Subsume`" otherwise.
+
+  Let <a grammar>`host-part`</a> be a `wildcard-host` if its first character is 
+  an U+002A ASTERISK character (*).
```
Comment (mikewest) [inline embedded/index.src.html]: Nit: Backtics around `*`.

File: embedded/index.src.html [hunk 5]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
...
+  <h4 id="subsume-se" algorithm>
+    Does |scheme-source / host-source expression| subsume another 
+    |scheme-source / host-source expression|?
+  </h4>
+
+  Given two <a grammar>scheme-source</a> / <a grammar>host-source</a> 
+  expressions |A| and |B|, this algorithm returns "`Subsumes`" if |A| subsumes
+  |B|, and returns "`Does Not Subsume`" otherwise.
+
+  Let <a grammar>`host-part`</a> be a `wildcard-host` if its first character is
```
Comment (mikewest) [inline embedded/index.src.html]: It's probably better to write this as "A <a grammar>`host-part`</a> expression is a <dfn>wildcard host</dfn> if its first character is ...". You can then reference it later as "If |A| is a <a>wildcard host</a> ..."

File: embedded/index.src.html [hunk 6]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
...
+    |scheme-source / host-source expression|?
+  </h4>
+
+  Given two <a grammar>scheme-source</a> / <a grammar>host-source</a> 
+  expressions |A| and |B|, this algorithm returns "`Subsumes`" if |A| subsumes
+  |B|, and returns "`Does Not Subsume`" otherwise.
+
+  Let <a grammar>`host-part`</a> be a `wildcard-host` if its first character is 
+  an U+002A ASTERISK character (*).
+  Let <a grammar>`port-part`</a> be a `wildcard-port` if it is equal to "*".
```
Comment (mikewest) [inline embedded/index.src.html]: Ditto.

File: embedded/index.src.html [hunk 7]
```diff
@@ -414,6 +414,60 @@ <h3 id="subsume-policy" algorithm>
...
+
+  1.  If [[csp3#match-schemes]] returns "`Does Not Match`" given |A|'s 
+      <a grammar>`scheme-part`</a> and |B|'s <a grammar>`scheme-part`</a>, return
```
Comment (mikewest) [inline embedded/index.src.html]: "... or `null`...", etc.

**Comment by Sun77789** [comment]
@mikewest Updated. Thank you for all the reviews!

File: embedded/index.src.html [hunk 8]
```diff
@@ -414,6 +414,79 @@ <h3 id="subsume-policy" algorithm>
...
+  Given two <a>source expressions</a> |A| and |B|, this algorithm returns 
+  "`Subsumes`" if |A| subsumes |B|, and returns "`Does Not Subsume`" otherwise.
+
+  Let <a>source expression</a> be a `wildcard host` if the first character of 
+  <a>source expressions</a>'s <a grammar>`host-part`</a> is an U+002A ASTERISK 
+  character (`*`).
+  Let <a>source expression</a> be a `wildcard port` if 
+  <a grammar>`port-part`</a> of <a>source expressions</a> is equal to "*".
+
+  1.  If |A| and |B| match either <a grammar>host-source</a> or
```
Comment (mikewest) [inline embedded/index.src.html]: "If _both_ |A| and |B| ...", right?

File: embedded/index.src.html [hunk 9]
```diff
@@ -414,6 +414,79 @@ <h3 id="subsume-policy" algorithm>
...
+  </h4>
+
+  Given two <a>source expressions</a> |A| and |B|, this algorithm returns 
+  "`Subsumes`" if |A| subsumes |B|, and returns "`Does Not Subsume`" otherwise.
+
+  Let <a>source expression</a> be a `wildcard host` if the first character of 
+  <a>source expressions</a>'s <a grammar>`host-part`</a> is an U+002A ASTERISK 
+  character (`*`).
+  Let <a>source expression</a> be a `wildcard port` if 
+  <a grammar>`port-part`</a> of <a>source expressions</a> is equal to "*".
```
Comment (mikewest) [inline embedded/index.src.html]: I'd still like you to rewrite these as described in https://github.com/w3c/webappsec-csp/pull/138#discussion_r86748255. Basically, turn them into `<dfn>` blocks:

"`A <a>source expression</a> has a <dfn>wildcard host</dfn> if ..." and "`If |A| has a <a>wildcard host</a>...`".

**Comment by Sun77789** [comment]
Squashing to one commit after latest updates

File: embedded/index.src.html [hunk 10]
```diff
@@ -414,6 +414,79 @@ <h3 id="subsume-policy" algorithm>
...
+
+      6.  If [[csp3#match-paths]] returns "`Does Not Match`" given |A|'s 
+          <a grammar>`path-part`</a> (or `null` if |A| does not contain a 
+          <a grammar>`path-part`</a>) and |B|'s <a grammar>`path-part`</a> (or 
+          `null` if |B| does not contain a <a grammar>`path-part`</a>), return 
+          "`Does Not Subsume`".
+
+      7.  Return "`Subsumes`".
+
+  2.  If both |A| and |B| both match the <a grammar>keyword-source</a> grammar:
```
Comment (mikewest) [inline embedded/index.src.html]: Too many "both"s. Pick one, drop the other, here and below. :)

Comment (Sun77789) [inline embedded/index.src.html]: Ooops! Updated!

**Pull request merged**